### ISSUE TITLE
Feature Request: Fixed Pricing Based on Day of Week

### ISSUE THREAD

**Issue by jakest** [MattJeanes/TeslaMateAgile#51]
My power plan is as follows:

Monday - Friday
- 00:00-08:00 low tier
- 08:00-10:00 high tier
- 10:00-15:00 low tier
- 15:00-20:00 high tier
- 20:00-00:00 low tier

Saturday - Sunday
- 00:00 - 23:59 low tier

When I charge during the weekend, the costs are calculated at a high tier rate if charging during one of the high tier times (only applicable on weekdays) as TeslaMateAgile can’t currently be setup with pricing tiers dependent upon the day of the week. 

It would be amazing to be able to set up Fixed Pricing based on the day of the week to account for that.

**Comment by MattJeanes** [comment]
Hello! Is this quite a common thing to have it vary across different days of the week? I've heard of it varying across the year before e.g. summer and winter pricing but not really across a week before.

If you have Home Assistant I'd recommend using that to set your prices as dynamically as you like, details in the readme of how to do that. If not, I could try to add something for this but I'm not sure when I'll be able to get around to it. Contributions welcome of course!

**Comment by jakest** [comment]
It does seem to be fairly common, at least in the mountain west states. The
pricing also varies across the year (summer vs winter) similar to what
you’ve mentioned, so I have a reminder to adjust the config when those
hours adjust.

The full day on the weekend is always in the lower tier pricing.

I don’t have Home Assistant currently, but maybe it’s time to check that
out.

**Comment by jfg31113** [comment]
Here in Spain we also have diferent plans for weekdays and weekends:

Monday - Friday

00:00-08:00 low tier
08:00-10:00 mid tier
10:00-14:00 high tier
14:00-18:00 mid tier
18::00-22:00 high tier
22:00-00:00 mid tier

Saturday - Sunday

00:00 - 23:59 low tier

**Comment by MattJeanes** [comment]
Hey @jakest / @jfg31113 - I have made a first version of this and I would love if you could help me test it.

I've just released test version [v1.17.0-rc.1](https://github.com/MattJeanes/TeslaMateAgile/releases/edit/v1.17.0-rc.1) which contains a new `FixedPriceWeekly` provider which should do what you want.

I have tested various different scenarios but I can't promise that it'll work perfectly, so I invite you to give it a go and try and break it, and most importantly let me know what you think!

See the [documentation](https://github.com/MattJeanes/TeslaMateAgile#fixed-price-weekly) for how to configure the new provider.

**Comment by jakest** [comment]
@MattJeanes I'm getting the following log error with the following config. 
[teslamate-teslamateagile-1.csv](https://github.com/user-attachments/files/17636574/teslamate-teslamateagile-1.csv)

```      System.ArgumentException: price (Parameter 'Failed to parse fixed price: Sat-Sun=0.052004') ```

```  teslamateagile:
    image: mattjeanes/teslamateagile:latest
    restart: always
    environment:
      - DATABASE_USER=XXXX
      - DATABASE_PASS=XXXX
      - DATABASE_NAME=XXXX
      - DATABASE_HOST=XXXX
      - TeslaMate__UpdateIntervalSeconds=3600
      - TeslaMate__GeofenceId=1
      - TeslaMate__EnergyProvider=FixedPrice
      - FixedPrice__TimeZone=America/Denver
      - FixedPrice__Prices__0=Mon-Fri=00:00-08:00=0.052004
      - FixedPrice__Prices__1=Mon-Fri=08:00-10:00=0.253532
      - FixedPrice__Prices__2=Mon-Fri=10:00-15:00=0.052004
      - FixedPrice__Prices__3=Mon-Fri=15:00-20:00=0.253532
      - FixedPrice__Prices__4=Mon-Fri=20:00-00:00=0.052004
      - FixedPrice__Prices__5=Sat,Sun=0.052004```

I've tried `Sat-Sun` as well. I've also tried `Sat-Sun=00:00-00:00=0.052004`. What should I be using for the same price the full weekend?

**Comment by MattJeanes** [comment]
Ah I see, make sure you are using the new `FixedPriceWeekly` in the `TeslaMate__EnergyProvider` setting 😄

**Comment by MattJeanes** [comment]
Oh, and all of the settings should be prefixed with `FixedPriceWeekly` instead of `FixedPrice` e.g. 

```yaml
      - TeslaMate__EnergyProvider=FixedPriceWeekly
      - FixedPriceWeekly__TimeZone=America/Denver
      - FixedPriceWeekly__Prices__0=Mon-Fri=00:00-08:00=0.052004
      - FixedPriceWeekly__Prices__1=Mon-Fri=08:00-10:00=0.253532
      - FixedPriceWeekly__Prices__2=Mon-Fri=10:00-15:00=0.052004
      - FixedPriceWeekly__Prices__3=Mon-Fri=15:00-20:00=0.253532
      - FixedPriceWeekly__Prices__4=Mon-Fri=20:00-00:00=0.052004
      - FixedPriceWeekly__Prices__5=Sat,Sun=0.052004
```

Apologies I needed to make it a brand new provider rather than extending the existing one hence the changes.

**Comment by jakest** [comment]
Ah! I didn't even notice, my mistake. Looks like I'm getting a new error now that is preventing the container from starting with this config.

```
  teslamateagile:
    image: mattjeanes/teslamateagile:latest
    restart: always
    environment:
      - DATABASE_USER=XXXX
      - DATABASE_PASS=XXXX
      - DATABASE_NAME=XXXX
      - DATABASE_HOST=XXXX
      - TeslaMate__UpdateIntervalSeconds=3600
      - TeslaMate__GeofenceId=1
      - TeslaMate__EnergyProvider=FixedPriceWeekly
      - FixedPriceWeekly__TimeZone=America/Denver
      - FixedPriceWeekly__Prices__0=00:00-08:00=0.052004
      - FixedPriceWeekly__Prices__1=08:00-10:00=0.253532
      - FixedPriceWeekly__Prices__2=10:00-15:00=0.052004
      - FixedPriceWeekly__Prices__3=15:00-20:00=0.253532
      - FixedPriceWeekly__Prices__4=20:00-00:00=0.052004
      - FixedPriceWeekly__Prices__5=Sat-Sun=0.052004
```
[teslamate-teslamateagile-1(2).csv](https://github.com/user-attachments/files/17638485/teslamate-teslamateagile-1.2.csv)

**Comment by MattJeanes** [comment]
Ah I didn't notice from your previous config but you will need to explicitly use the preview version as `latest` won't get you the preview ones 🙂

`image: mattjeanes/teslamateagile:latest` -> `image: mattjeanes/teslamateagile:v1.17.0-rc.1`

**Comment by jakest** [comment]
I totally missed that, my fault. Was going to look at the SHA hash, then got busy with work. Looks like we had our first run! Looks like I only have one error now.

```
date	stream	content
2024/11/05 15:22:32	stdout	      Waiting 3600 seconds until next update
2024/11/05 15:22:32	stdout	info: TeslaMateAgile.PriceService[0]
2024/11/05 15:22:32	stdout	         at TeslaMateAgile.PriceHelper.Update() in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 75
2024/11/05 15:22:32	stdout	         at TeslaMateAgile.PriceHelper.CalculateChargeCost(IEnumerable`1 charges) in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 98
2024/11/05 15:22:32	stdout	         at TeslaMateAgile.PriceHelper.CalculateDynamicChargeCost(IEnumerable`1 charges, DateTimeOffset minDate, DateTimeOffset maxDate) in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 108
2024/11/05 15:22:32	stdout	         at TeslaMateAgile.Services.FixedPriceWeeklyService.GetPriceData(DateTimeOffset from, DateTimeOffset to) in /src/TeslaMateAgile/Services/FixedPriceWeeklyService.cs:line 36
2024/11/05 15:22:32	stdout	         at TeslaMateAgile.Services.FixedPriceWeeklyService.GetPriceDataForDate(DateTime date) in /src/TeslaMateAgile/Services/FixedPriceWeeklyService.cs:line 124
2024/11/05 15:22:32	stdout	      System.Exception: Invalid fixed price data, prices are not continuous
2024/11/05 15:22:32	stdout	      Failed to calculate charging cost / energy for charging process 2610
2024/11/05 15:22:32	stdout	fail: TeslaMateAgile.PriceHelper[0]
2024/11/05 15:22:32	stdout	      Calculating cost for charges 11/05/2024 19:15:12 UTC - 11/05/2024 20:50:33 UTC
2024/11/05 15:22:32	stdout	info: TeslaMateAgile.PriceHelper[0]
2024/11/05 15:22:31	stdout	      Looking for finished charging processes with no cost set in the 'Home' geofence (id: 1)
2024/11/05 15:22:31	stdout	info: TeslaMateAgile.PriceHelper[0]
2024/11/05 15:22:26	stdout	      Content root path: /app/
2024/11/05 15:22:26	stdout	info: Microsoft.Hosting.Lifetime[0]
2024/11/05 15:22:26	stdout	      Hosting environment: Production
2024/11/05 15:22:26	stdout	info: Microsoft.Hosting.Lifetime[0]
2024/11/05 15:22:26	stdout	      Application started. Press Ctrl+C to shut down.
2024/11/05 15:22:26	stdout	info: Microsoft.Hosting.Lifetime[0]
2024/11/05 15:22:26	stdout	      Using energy provider FixedPriceWeekly
2024/11/05 15:22:26	stdout	info: TeslaMateAgile.PriceService[0]
2024/11/05 15:22:26	stdout	      Price service is starting
2024/11/05 15:22:26	stdout	info: TeslaMateAgile.PriceService[0] ```

It failed on the latest charge for some reason. Feels like we're so close.

![2024-11-05_15-27-04](https://github.com/user-attachments/assets/7e1d53cf-e72f-4c9d-b68e-54ace2ab4d17)

**Comment by jakest** [comment]
Actually, I just deleted a few more costs to see if it would update any of them. It looks like it's failing to update any entries that don't have a cost associated yet.

**Comment by MattJeanes** [comment]
Thanks, looking into that now - I can reproduce the error locally. It's erroring as part of a self-check in the internal part of the code, nothing wrong on your side this time!

**Comment by MattJeanes** [comment]
Okay, fixed! Looks like if your one of your time ranges crosses over exactly midnight it breaks! Please try the new version [v1.17.0-rc.2](https://github.com/MattJeanes/TeslaMateAgile/releases/tag/v1.17.0-rc.2) which resolves this problem.

```yaml
image: mattjeanes/teslamateagile:v1.17.0-rc.2
```

**Comment by jakest** [comment]
Look at that! The one I didn't have a price for succeeded. The one I deleted the pricing information for manually is still failing, but since I only deleted it to test, it's likely not an edge case worth fixing. But just in case, here are the logs for that one.

```
2024/11/05 15:51:41	stdout	info: TeslaMateAgile.PriceHelper[0]
2024/11/05 15:51:41	stdout	         at TeslaMateAgile.PriceHelper.Update() in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 75
2024/11/05 15:51:41	stdout	         at TeslaMateAgile.PriceHelper.CalculateChargeCost(IEnumerable`1 charges) in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 98
2024/11/05 15:51:41	stdout	         at TeslaMateAgile.PriceHelper.CalculateDynamicChargeCost(IEnumerable`1 charges, DateTimeOffset minDate, DateTimeOffset maxDate) in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 151
2024/11/05 15:51:41	stdout	      System.Exception: Charge calculation failed, pricing calculated for 0 / 54, likely missing price data
2024/11/05 15:51:41	stdout	      Failed to calculate charging cost / energy for charging process 2607
2024/11/05 15:51:41	stdout	fail: TeslaMateAgile.PriceHelper[0]
2024/11/05 15:51:41	stdout	      Calculating cost for charges 11/04/2024 01:22:55 UTC - 11/04/2024 01:40:58 UTC 
```

By the way, do you have a way to accept donations? I really appreciate your help with this!

**Comment by MattJeanes** [comment]
Thank you, that is another useful test case too. Apologies, this new provider is rather complex in it's logic so trying to squash as many bugs as I can in it before releasing it to everyone. I can reproduce this issue as well, investigating now!

**Comment by MattJeanes** [comment]
Okay, fixed that problem too. That one was caused by the timezone jumping into the previous day (relative to UTC) and not correctly retrieving the prices for the previous date. I love timezones 😁

The new version [v1.17.0-rc.3](https://github.com/MattJeanes/TeslaMateAgile/releases/tag/v1.17.0-rc.3) should fix this problem!

```yaml
image: mattjeanes/teslamateagile:v1.17.0-rc.3
```

As for donations, I've also just set up GitHub Sponsors if you want to contribute anything! I don't ask for donations and I do this work for free to give back to the community, but if you did want to I would appreciate your support a lot 😊

You can access GitHub Sponsors on the [TeslaMateAgile repository main page](https://github.com/MattJeanes/TeslaMateAgile) or on my [GitHub profile page](https://github.com/MattJeanes), there should be a 'Sponsor' button there now.

**Comment by jakest** [comment]
I didn't even consider that. Good find! That resolved the issue. I'll let you know if I see anything else over the next few weeks.

Thanks again for adding this!

**Comment by MattJeanes** [comment]
Awesome, yes please keep me posted if anything is acting weird or erroring I would love to hear about it.

Keep in mind that even if it works there's a chance the calculated cost might be wrong if my calculations are wrong, which is definitely a possibility 👀 so if anything looks weird or doesn't seem to add up please shout!

You're welcome, and thank you also for helping me test it, real world test scenarios are so much more useful than my fake ones I made for testing which already missed a couple of real issues as we found out 😁

I've also just seen your donation, thank you very much!! It means a lot that you value my work enough to do that ❤️

**Comment by MattJeanes** [comment]
Hey @jakest - how's it been going? All good still or seeing any issues?

**Comment by jakest** [comment]
It has been working perfectly! Thanks for checking in. I’ve had charges
during peak, off peak, on every day, and some that cross over. I think
every edge case is covered now.

Thanks again!

**Comment by MattJeanes** [comment]
Awesome! Thank you for testing that out - let me know if you spot anything in future of course!

I'll keep this issue open until I push the release out properly, just waiting on some feedback on a different change in this update first 🙂

**Comment by MattJeanes** [comment]
This has now been released in version [`v1.17.0`](https://github.com/MattJeanes/TeslaMateAgile/releases/tag/v1.17.0) 🙂
